docs/proposals/201812_thanos-remote-receive.md

```diff
@@ -0,0 +1,166 @@
+
+```
+                                  Soft tenant hashring
+                                 +-----------------------+
+                                 |                       |
++-----------------+              |  +-----------------+  |
+|                 |              |  |                 |  |
+|  Load Balancer  +-------+      |  | Thanos receiver |  |
+|                 |       |      |  |                 |  |
++-----------------+       |      |  +-----------------+  |
+                          |      |                       |
+                          |      |                       |
+                          |      |  +-----------------+  |
+                          |      |  |                 |  |
+                          +-------->+ Thanos receiver +-----------+
+                                 |  |                 |  |        |
+                                 |  +-----------------+  |        |
+                                 |                       |        |
+                                 +-----------------------+        |
+                                                                  |
+                                   Hard Tenant A hashring         |
+                                 +-----------------------+        |
+                                 |                       |        |
+                                 |  +-----------------+  |        |
+                                 |  |                 |  |        |
+                                 |  | Thanos receiver +<----------+
+                                 |  |                 |  |        |
+                                 |  +-----------------+  |        |
+                                 |                       |        |
+                                 |                       |        |
+                                 |  +-----------------+  |        |
+                                 |  |                 |  |        |
+                                 |  | Thanos receiver +<----------+
+                                 |  |                 |  |
+                                 |  +-----------------+  |
+                                 |                       |
+                                 +-----------------------+
+```
+
+The intention is that the load balancer can just distribute requests randomly to all Thanos receivers independent of the tenant. Should this turn out to cause problems, either an additional “distribution” layer can be inserted that performs this routing or a load balancer per hard tenant hashring can be introduced. The distribution layer would have the advantage of being able to keep a single endpoint for configuration of Prometheus servers for remote write, which could be convenient for users.
+
+### Rollout/scaling/failure of receiver nodes
```

Comment:
Do I understand correctly that on each upgrade of a node or scaling up or down of a node we will flush the tsdb to block of a single node to object storage.
On changing of soft to hard tenant node (or vise versa) we need to flush blocks in all nodes in the effected rings. 
In the scaling scenarios does it not mean the tsdb data for a tenant may end up on a different node? And therefore should flush the whole ring effected on any scaling?
I take it if any traffic coming to a `flushing` node will get a `503` immediately.

Reply:
There are a couple of statements in there I'll answer them individually. Let me know if there is anything unclear (also if any of this needs to be more clear in the proposal).

> I take it if any traffic coming to a flushing node will get a 503 immediately.

 Correct, that's why intermediate downtime for ingestion is acceptable. Prometheus remote write treats 503 as temporary failures and continues do retry until the remote write receiving end responds again.

> In the scaling scenarios does it not mean the tsdb data for a tenant may end up on a different node? And therefore should flush the whole ring effected on any scaling?

Yes flushing the whole ring will be the initial implementation. With consistent hashing only 1/n data needs moving, so it can be optimized to that.

> On changing of soft to hard tenant node (or vise versa) we need to flush blocks in all nodes in the effected rings.

Correct.